File: src/main/java/fr/marstech/mtlinkspray/service/ReportAbuseServiceImpl.java
package fr.marstech.mtlinkspray.service;

import fr.marstech.mtlinkspray.entity.AbuseReportEntity;
import fr.marstech.mtlinkspray.entity.HistoryItem;
import fr.marstech.mtlinkspray.repository.AbuseReportRepository;
import jakarta.servlet.http.HttpServletRequest;
import org.springframework.stereotype.Service;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Map;
import java.util.UUID;

@Service
public class ReportAbuseServiceImpl implements ReportAbuseService {

    private final AbuseReportRepository abuseReportRepository;
    private final MailSenderService mailSenderService;

    public ReportAbuseServiceImpl(
            AbuseReportRepository abuseReportRepository, MailSenderService mailSenderService) {
        this.abuseReportRepository = abuseReportRepository;
        this.mailSenderService = mailSenderService;
    }

    @Override
    public void reportAbuse(String inputAbuseDecsription, HttpServletRequest httpServletRequest) {

        String action = "reportAbuse";
        String report = "%s:%s".formatted(action, inputAbuseDecsription);

        HistoryItem historyItem =
                new HistoryItem(httpServletRequest.getRemoteAddr(), LocalDateTime.now(), action);

        AbuseReportEntity abuseReportEntity =
                new AbuseReportEntity(
                        UUID.randomUUID().toString(),
                        LocalDateTime.now(),
                        null,
                        true,
                        null,
                        Map.of(),
                        new HistoryItem(),
                        List.of());
        abuseReportEntity.setDescription(report);
        abuseReportEntity.setAuthor(historyItem);
        abuseReportEntity = abuseReportRepository.save(abuseReportEntity);

        mailSenderService.sendMail(" MarsTech Link-Spray - Abuse report", abuseReportEntity.toString());
    }
}

File: src/main/kotlin/fr/marstech/mtlinkspray/service/ReportAbuseServiceImpl.kt
package fr.marstech.mtlinkspray.service

import fr.marstech.mtlinkspray.entity.AbuseReportEntity
import fr.marstech.mtlinkspray.entity.HistoryItem
import fr.marstech.mtlinkspray.repository.AbuseReportRepository
import jakarta.servlet.http.HttpServletRequest
import org.springframework.stereotype.Service
import java.time.LocalDateTime
import java.util.*
import java.util.Map

@Service
class ReportAbuseServiceImpl(
    private val abuseReportRepository: AbuseReportRepository,
    private val mailSenderService: MailSenderService
) : ReportAbuseService {

    override fun reportAbuse(
        inputAbuseDecsription: String, httpServletRequest: HttpServletRequest
    ) = "reportAbuse".let { it ->
        AbuseReportEntity(
            id = UUID.randomUUID().toString(),
            creationDate = LocalDateTime.now(),
            expiresAt = null,
            isEnabled = true,
            description = "$it:$inputAbuseDecsription",
            metadata = Map.of<String, String>(),
            author = HistoryItem(
                ipAddress = httpServletRequest.remoteAddr,
                dateTime = LocalDateTime.now(),
                action = it
            )
        ).let { abuseReportRepository.save(it) }
            .let {
                mailSenderService.sendMail(
                    subject = " MarsTech Link-Spray - Abuse report",
                    body = it.toString()
                )
            }

    }
}
